src/main/python/deep_qa/models/reading_comprehension/bidirectional_attention.py

```diff
@@ -0,0 +1,197 @@
+
+        # Shape: (batch_size, num_passage_words, num_question_words), normalized over question
+        # words for each passage word.
+        passage_question_attention = MaskedSoftmax()(passage_question_similarity)
+        # Shape: (batch_size, num_passage_words, embedding_size * 2)
+        passage_question_vectors = WeightedSum()([encoded_question, passage_question_attention])
+
+        # Min's paper finds, for each document word, the most similar question word to it, and
+        # computes a single attention over the whole document using these max similarities.
+        # Shape: (batch_size, num_passage_words)
+        question_passage_similarity = Max(axis=-1)(passage_question_similarity)
+        # Shape: (batch_size, num_passage_words)
+        question_passage_attention = MaskedSoftmax()(question_passage_similarity)
+        # Shape: (batch_size, embedding_size * 2)
+        question_passage_vector = WeightedSum()([encoded_passage, question_passage_attention])
+
+        # Then he repeats this question/passage vector for every word in the passage, and uses it
+        # as an additional input to the hidden layers above.
+        repeat_layer = Repeat(axis=1, repetitions=self.num_passage_words)
+        # Shape: (batch_size, num_passage_words, embedding_size * 2)
+        tiled_question_passage_vector = repeat_layer(question_passage_vector)
+
+        # Shape: (batch_size, num_passage_words, embedding_size * 8)
+        final_merged_passage = ComplexConcat(combination='1,2,1*2,1*3')([encoded_passage,
+                                                                         passage_question_vectors,
+                                                                         tiled_question_passage_vector])
+
+        # PART 3:
+        # Having computed a combined representation of the document that includes attended question
+        # vectors, we'll pass this through a few more bi-directional encoder layers, then predict
+        # the span_begin word.  Hard to find a good name for this; Min calls this part of the
+        # network the "modeling layer", so we'll call this the `modeled_passage`.
+        modeled_passage = final_merged_passage
+        for i in range(self.num_hidden_seq2seq_layers):
+            # TODO(matt): fix the TextTrainer encoder API to take named encoders, make this easier.
+            hidden_layer = self._get_new_seq2seq_encoder(self._get_seq2seq_params(),
+                                                         name="hidden_seq2seq_{}".format(i))
+            modeled_passage = hidden_layer(modeled_passage)
+
+        # To predict the span word, we pass the merged representation through a Dense layer without
+        # output size 1 (basically a dot product of a vector of weights and the passage vectors),
+        # then do a softmax to get a position.
+        span_begin_input = merge([final_merged_passage, modeled_passage], mode='concat')
```

Comment:
is there a reason why it's `merge` as opposed to `Merge`? The keras API seems to suggest the class name is `Merge` as well... what am I missing?